fish-rust/src/common.rs

```diff
@@ -1,13 +1,1707 @@
+///
+/// This function encodes illegal character sequences in a reversible way using the private use
+/// area.
+pub fn str2wcstring(inp: &[u8]) -> WString {
+    if inp.is_empty() {
+        return WString::new();
+    }
+
+    let mut result = WString::new();
+    result.reserve(inp.len());
+    let mut pos = 0;
+    let mut state = zero_mbstate();
+    while pos < inp.len() {
+        // Append any initial sequence of ascii characters.
+        // Note we do not support character sets which are not supersets of ASCII.
+        let ascii_prefix_length = count_ascii_prefix(&inp[pos..]);
+        result.push_str(std::str::from_utf8(&inp[pos..pos + ascii_prefix_length]).unwrap());
+        pos += ascii_prefix_length;
+        assert!(pos <= inp.len(), "Position overflowed length");
+        if pos == inp.len() {
+            break;
+        }
+
+        // We have found a non-ASCII character.
+        let mut ret = 0;
+        let mut wc = '\0';
+
+        let use_encode_direct = if inp[pos] & 0xF8 == 0xF8 {
+            // Protect against broken mbrtowc() implementations which attempt to encode UTF-8
+            // sequences longer than four bytes (e.g., OS X Snow Leopard).
+            // TODO This used to be optimized away at compile time.
+            true
+        } else {
+            let mut wc_wchar = wc as libc::wchar_t;
+            ret = unsafe {
+                mbrtowc(
+                    std::ptr::addr_of_mut!(wc_wchar),
+                    std::ptr::addr_of!(inp[pos]).cast(),
+                    inp.len() - pos,
+                    std::ptr::addr_of_mut!(state),
+                )
+            };
+            wc = char::from_u32(wc_wchar as u32).unwrap();
+            // Determine whether to encode this character with our crazy scheme.
+            (wc >= ENCODE_DIRECT_BASE && wc < ENCODE_DIRECT_END)
+            ||
+            wc == INTERNAL_SEPARATOR
+            ||
+            // Incomplete sequence.
+            ret == 0_usize.wrapping_sub(2)
+            ||
+            // Invalid data.
+            ret == 0_usize.wrapping_sub(1)
+            ||
+            // Other error codes? Terrifying, should never happen.
+            ret > inp.len() - pos
+        };
+
+        if use_encode_direct {
+            wc = wchar_literal_byte(inp[pos]);
+            result.push(wc);
+            pos += 1;
+            state = zero_mbstate();
+        } else if ret == 0 {
+            // embedded null byte!
+            result.push('\0');
+            pos += 1;
+            state = zero_mbstate();
+        } else {
+            // normal case
+            result.push(wc);
+            pos += ret;
+        }
+    }
+    result
+}
+
+/// Returns a newly allocated multibyte character string equivalent of the specified wide character
+/// string.
+///
+/// This function decodes illegal character sequences in a reversible way using the private use
+/// area.
+pub fn wcs2string(input: &wstr) -> Vec<u8> {
```

Comment:
Make it two functions?

```rust
pub fn wcs2string(input: &wstr) -> Vec<u8>; // Or OsString
pub fn wcs2zstring(input: &wstr) -> CString;
```

Most callers will use `wcs2zstring`. Maybe we should rename `wcs2string`.
Could also do the rename on the C++ side if that helps.

Reply:
Two functions makes some sense to me, since today interior nul bytes can just data to be dropped. Example:

    echo text > abc\0def.txt

will populate a file `abc`. We might want to emit an error for that case, which requires us to know if the string is going to be used as a path or not.

Reply:
@krobelus IMHO `OsString` would be preferable over `Vec<u8>` since it makes the intention and usage clearer.

-----

> I think we need a real nul-terminated string type that allows interior nuls. I'm not sure what it should be. What do you think?

@ridiculousfish what would a nul-terminated string that supports interior nuls look like? You'd necessarily have to use the count/len on one end to support that interior nul meaning the terminating nul would be hanging out there at the end "just because" (you couldn't rely on it for anything) since either both sides of the interface use the count (rendering the nul-termination moot) or one side uses the len but the other side uses the nul termination, meaning the string gets truncated at first nul.

IMHO we shouldn't use nul-termination for anything except where the (legacy/system) callee requires it. In that case, what would make the most sense is to simply have either a type that cannot contain interior nulls (the setter/new takes a regular `&str`/`&wstr` but errors out or returns `None` if it contained an interior nul) or use a conversion function at the call site that copies to a nul-terminated buffer but similarly throws/panics/fails if the source string contained an interior `nul`.

Reply:
A nul-terminated string that supports interior nuls looks like `std::string`. If you pass it to a syscall like `open` which expects nul-termination then it will truncate your filename; but if you pass it to `write` which has pointer/length pair, it will happily write the zero byte, as it should. So it depends.

What you propose is the two-function approach: interior nuls for some cases (a binary file stream), prohibit them for others (a file redirection), and the call sites have to decide which to use. If that's achievable then I'm for it. It may also be too difficult - we use `wcs2string` a lot, and maybe it's too difficult to tease apart these two use cases, so we should just have some Rust analog to `std::string`.

I want to defer to krobelus here; we should aim to match the existing behavior and we can defer any refactoring to later.

Reply:
> IMHO OsString would be preferable over Vec<u8> since it makes the intention and usage clearer.

Yes but I'm not sure where to draw the line. In the strictest interpretation, we would change almost every use of `Vec<u8>` to `OsString` (and `&[u8]` to `OsStr`) because we almost never deal with binary data.
OsString is a more limited API, for example it doesn't have `pop()`.
Also, `OsString::push()` takes a `OsSt`, we cannot push a char directly.
This suggests that we should still use `Vec<u8>` when manipulating a string and convert to `OsStr` later.
There is another problem: `OsStr` does not support slice indexing.

I'm positive that it's worth trying to adopt `OsStr` but it needs more work and thought.
While we use `Vec<u8>`, we do need to convert to `OsStr` whenever we calling something that takes `AsRef<Path>` but that's a small price to pay.

Reply:
The point of `OsStr` is basically that you can't do anything with it directly, since its representation is platform-dependent. That's fine for "take this data from the kernel, put it back in the kernel somewhere else", but less useful for a shell that actually has to look at the contents and manipulate them. I think for the majority of use cases in fish, `OsStr` is the wrong abstraction.

Reply:
Agree with Xiretza to avoid OsString.

Looking at the call sites, I think the two function approach should be very doable. The great majority of `wcs2string` calls are in preparation for a system call (e.g. `wunlink`); these can use `std::ffi::CString` which is what it's there for.

The only important exception is `io_buffer_t` which holds a buffer, e.g. the output of a command when executing a command substitution. This holds a `std::string`, needs to support interior nuls, but does NOT need to be nul-terminated. A `Vec<u8>` would be totally fine here.

Reply:
> It may also be too difficult - we use wcs2string a lot, and maybe it's too difficult to tease apart these two use cases, so we should just have some Rust analog to std::string.

From the perspective of simply finishing the port, I fully agree. 

However I would say silently truncating *some* strings at the first null but not others is a bug - the operation should certainly fail rather than half-succeed.

Reply:
CString does not allow interior nulls so this check is easy to implement.
We'd need to make `wcs2zstring` return an error instead of truncating the string.
Probably not worth the effort though.

Reply:
If we get idiomatic rust error handling going, it would be as easy as adding a question mark at any call site and defining an `Into` for the error that automatically sets the error message to something appropriate. But probably best left for later.